drivers/ps2/ps2_npcx_controller.c

```diff
@@ -0,0 +1,354 @@
+	 * (due to Shift Mechanism is reset)
+	 */
+	return (IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT) |
+		IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_EOT)) ?
+		       -EBUSY :
+		       0;
+}
+
+int ps2_npcx_ctrl_write(const struct device *dev, uint8_t channel_id,
+			uint8_t value)
+{
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	int i = 0;
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
+	}
+
+	if (!(BIT(NPCX_PSOSIG_CLK(channel_id)) & data->channel_enabled_mask)) {
+		LOG_ERR("channel %d is not enabled", channel_id);
+		return -EINVAL;
+	}
+
+	k_sem_take(&data->lock, K_FOREVER);
+
+	while (ps2_npcx_ctrl_bus_busy(dev) && (i < PS2_RETRY_COUNT)) {
+		k_busy_wait(50);
+		i++;
+	}
+
+	if (unlikely(i == PS2_RETRY_COUNT)) {
+		LOG_ERR("PS2 write attempt timed out");
+		goto timeout_invalid;
+	}
+
+	/* Set PS/2 in transmit mode */
+	inst->PSCON |= BIT(NPCX_PSCON_XMT);
+	/* Enable Start Of Transaction interrupt */
+	inst->PSIEN |= BIT(NPCX_PSIEN_SOTIE);
+
+	/* Reset the shift mechanism */
+	inst->PSOSIG = NPCX_PS2_SHIFT_MECH_RESET;
+	/* Inhibit communication should last at least 100 micro-seconds */
+	k_busy_wait(100);
+
+	/* Write the data to be transmitted */
+	inst->PSDAT = value;
+	/* Apply the Request-to-send */
+	inst->PSOSIG &= ~BIT(NPCX_PSOSIG_WDAT(channel_id));
+	inst->PSOSIG |= BIT(NPCX_PSOSIG_CLK(channel_id));
+	if (k_sem_take(&data->tx_sync_sem, PS2_TRANSACTION_TIMEOUT) != 0) {
+		irq_disable(DT_INST_IRQN(0));
+		LOG_ERR("PS/2 Tx timeout");
+		/* Reset the shift mechanism */
+		inst->PSOSIG = NPCX_PS2_SHIFT_MECH_RESET;
+		/* Change the PS/2 module to receive mode */
+		inst->PSCON &= ~BIT(NPCX_PSCON_XMT);
+		/*
+		 * Restore the channel back to enable according to
+		 * channel_enabled_mask.
+		 */
+		inst->PSOSIG |= data->channel_enabled_mask;
+		irq_enable(DT_INST_IRQN(0));
+		goto timeout_invalid;
+	}
+
+	k_sem_give(&data->lock);
+	return 0;
+
+timeout_invalid:
+	k_sem_give(&data->lock);
+	return -ETIMEDOUT;
+}
+
+static int ps2_npcx_ctrl_is_rx_error(const struct device *dev)
+{
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	uint8_t status;
+
+	status = inst->PSTAT & (BIT(NPCX_PSTAT_PERR) | BIT(NPCX_PSTAT_RFERR));
+	if (status) {
+		if (status & BIT(NPCX_PSTAT_PERR))
+			LOG_ERR("RX parity error");
+		if (status & BIT(NPCX_PSTAT_RFERR))
+			LOG_ERR("RX Frame error");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void ps2_npcx_ctrl_isr(const struct device *dev)
+{
+	uint8_t active_ch, mask;
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	/*
+	 * ACH = 1 : CHannel 0
+	 * ACH = 2 : CHannel 1
+	 * ACH = 4 : CHannel 2
+	 * ACH = 5 : CHannel 3
+	 */
+	active_ch = GET_FIELD(inst->PSTAT, NPCX_PSTAT_ACH);
+	active_ch = active_ch > 2 ? (active_ch - 2) : (active_ch - 1);
+	LOG_DBG("ACH: %d\n", active_ch);
```

Comment:
Do you need to range-check active_ch?

Reply:
The active channel number is reported from the hardware module.
It should not report incorrect channel No. unless there is a bug in the hardware.
So I don't check the range of it.